keps/sig-scheduling/5278-nominated-node-name-for-expectation/README.md

```diff
@@ -0,0 +1,983 @@
+speeding up the filter phase.
+
+### Retain the scheduling decision
+
+At the binding cycle (e.g., PreBind), some plugins could handle something (e.g., volumes, devices) based on the pod's scheduling result.
+
+If the scheduler restarts while it's handling some pods at binding cycles,
+kube-scheduler could decide to schedule a pod to a different node. 
+If we can keep where the pod was going to go at `NominatedNodeName`, the scheduler likely picks up the same node, and the PreBind plugins can restart their work from where they were before the restart.
+
+### Goals
+
+- The scheduler use `NominatedNodeName` to express where the pod is going to go before actually bound them.
+- Make sure external components can use `NominatedNodeName` to express where they prefer the pod is going to.
+  - Probably, you can do this with a today's scheduler as well. This proposal wants to discuss/make sure if it actually works, and then add tests etc.
+
+### Non-Goals
+
+- Extenral components can enforce the scheduler to pick up a specific node via `NominatedNodeName`.
+  - `NominatedNodeName` is just a hint for scheduler and doesn't represent a hard requirement
+
+## Proposal
+
+### User Stories (Optional)
+
+Here is the all use cases of NominatedNodeNames that we're taking into consideration:
+- The scheduler puts it after the preemption (already implemented)
+- The scheduler puts it at the beginning of binding cycles (only if the binding cycles invole PreBind phase)
+- The cluster autoscaler puts it after creating a new node for pending pod(s) so that the scheduler can find a place faster when the node is created.
+- Kueue uses it to determine a prefered node for the pod based on their internal calculation (Topology aware scheduling etc)
+
+(Possibly, our future initiative around the workload scheduling (including gang scheduling) can also utilize it,
+but we don't discuss it here because it's not yet concreted at all.)
+
+#### Story 1: Prevent inappropriate scale downs by Cluster Autoscaler 
+
+The scheduler starts to expose where the pod is going to with `NominatedNodeName` at the beginning of binding cycles.
+And, the cluster autoscaler takes `NominatedNodeName` into consideration when calculating which nodes they delete.
+
+It helps the scenarios where the binding cycles take time, for example, VolumeBinding plugin takes time at PreBind extension point.
+
+#### Story 2: Cluster Autoscaler specifies `NominatedNodeName` to indicate where pods can go after new nodes are created/registered
+
+Usually, the scheduler scans all the nodes in the cluster when scheduling pods.
+
+When the cluster autoscaler creates instances for pending pods, it calculate which new node might get which pending pod.
+If they can put `NominatedNodeName` based on those calculation, it could tell the scheduler that the node can probably picked up for the pod's scheduling,
```

Comment:
But scheduler can still process another pod earlier and land it on the new node, correct? NNN doesn't eliminate this race condition unless scheduler somehow reserves the capacity for pods with NNN set.

Reply:
> unless scheduler somehow reserves the capacity for pods with NNN set.

It actually does. When NNN is set to pending pod, the scheduler reserves the place for this pod on the node, and this reservation can only be ignored by higher priority pods.